Sources/SwiftProtobuf/AnyMessageStorage.swift

```diff
@@ -162,35 +162,81 @@ internal class AnyMessageStorage {
+    visitor.visitAnyVerbose(value: message, typeURL: url)
+  }
+
+  // Specialized traverse for writing out a Text form of the Any.
+  // This prefers the more-legible "verbose" format if it can
+  // use it, otherwise will fall back to simpler forms.
+  private func textTraverse(visitor: inout TextFormatEncodingVisitor) {
+    if let msg = _message {
+      emitVerboseTextForm(visitor: &visitor, message: msg, typeURL: _typeURL)
+    } else if let valueData = _valueData {
+      if let messageType = Google_Protobuf_Any.messageType(forTypeURL: _typeURL) {
+        // If we can decode it, we can write the readable verbose form:
+        do {
+          let m = try messageType.init(serializedData: valueData)
+          emitVerboseTextForm(visitor: &visitor, message: m, typeURL: _typeURL)
+          return
+        } catch {
+          // Fall through to just print the type and raw binary data
+        }
+      }
+      if !_typeURL.isEmpty {
+        try! visitor.visitSingularStringField(value: _typeURL, fieldNumber: 1)
+      }
+      if !valueData.isEmpty {
+        try! visitor.visitSingularBytesField(value: valueData, fieldNumber: 2)
+      }
+    } else if let contentJSON = _contentJSON {
+      // Build a readable form of the JSON:
+      let asciiOpenCurlyBracket = UInt8(ascii: "{")
+      let asciiCloseCurlyBracket = UInt8(ascii: "}")
+      var contentJSONAsObject = Data(bytes: [asciiOpenCurlyBracket])
+      contentJSONAsObject.append(contentJSON)
+      contentJSONAsObject.append(asciiCloseCurlyBracket)
+      // If we can decode it, we can write the readable verbose form:
+      if let messageType = Google_Protobuf_Any.messageType(forTypeURL: _typeURL) {
+        var any = Google_Protobuf_Any()
+        any.typeURL = _typeURL
+        any._storage._contentJSON = contentJSON
+        any._storage._valueData = nil
+        do {
+          let m = try messageType.init(unpackingAny: any)
+          emitVerboseTextForm(visitor: &visitor, message: m, typeURL: _typeURL)
+          return
+        } catch {
+          // Fall through to just print the raw JSON data
+        }
+      }
+      if !_typeURL.isEmpty {
+        try! visitor.visitSingularStringField(value: _typeURL, fieldNumber: 1)
+      }
+      visitor.visitAnyJSONDataField(value: contentJSONAsObject)
+    } else if !_typeURL.isEmpty {
+      try! visitor.visitSingularStringField(value: _typeURL, fieldNumber: 1)
     }
   }
 
-  // Called before the message is traversed to do any error preflights.
-  // Since traverse() will use _value, this is our chance to throw
-  // when _value can't.
-  func preTraverse() throws {
+  // Specialized traverse() implementation for Any
+  // This does various checks depending on the output format
+  // and contents, then traverses the storage fields.
+  func traverse<V: Visitor>(visitor: inout V) throws {
+    // Encoding to Text format for debugging is special:
+    if visitor is TextFormatEncodingVisitor {
```

Comment:
Thinking a little move about this, having to do this check within Any proves that our visitor api can't be used to write the TextFormat support.  Checkout outside is what someone else would likely have to do.  (yes, there's the catch that they any doesn't expose everything to do this, but it still provides some what of a comparison)